base/compiler/optimize.jl

```diff
@@ -254,16 +254,31 @@ function stmt_effect_flags(lattice::AbstractLattice, @nospecialize(stmt), @nospe
             nothrow = is_nothrow(effects)
             return (consistent, effect_free & nothrow, nothrow)
         elseif head === :new
-            typ = argextype(args[1], src)
+            atyp = argextype(args[1], src)
             # `Expr(:new)` of unknown type could raise arbitrary TypeError.
-            typ, isexact = instanceof_tfunc(typ)
-            isexact || return (false, false, false)
-            isconcretedispatch(typ) || return (false, false, false)
+            typ, isexact = instanceof_tfunc(atyp)
+            if !isexact
+                atyp = unwrap_unionall(widenconst(atyp))
+                if isType(atyp) && isTypeDataType(atyp.parameters[1])
```

Comment:
I thought at some point we had to deal with this check being insufficient and incomplete due to the existence of free-type-vars, for example if the user called:
```
julia> Val(Val.body);
```
which I thought used to be an issue for `convert` and `Fix1` calls (particularly those implied by fieldtypes)

But perhaps we don't handle that case anyways, since that call segfaults, and this call poisons the dispatch cache permanently:
```
julia> Base.Fix1(Val.body, 1)
ERROR: UndefVarError: F not defined
```

Reply:
(see test/core.jl for `TypeError("new", ...`)